strawberryfields/configuration.py

```diff
@@ -15,28 +15,136 @@
+    ``dict.update`` method, modified to recursively update
+    keys with nested dictionaries.
+    """
+    for key, value in overrides.items():
+        if isinstance(value, collections.Mapping) and value:
+            # Override value is a non-empty dictionary.
+            # Update the source key with the override dictionary.
+            returned = _deep_update(source.get(key, {}), value)
+            source[key] = returned
+        elif value != {}:
+            # Override value is not an empty dictionary.
+            source[key] = overrides[key]
+    return source
+
+
+def _generate_config(config_spec, **kwargs):
+    """Generates a configuration, given a Strawberry Fields configuration
+    specification.
+
+    See :attr:`~.DEFAULT_CONFIG_SPEC` for an example of a valid configuration
+    specification.
+
+    Optional keyword arguments may be provided to override default values
+    in the cofiguration specification. If the provided override values
+    do not match the expected type defined in the configuration spec,
+    a ``ConfigurationError`` is raised.
+
+    **Example**
+
+    >>> _generate_config(DEFAULT_CONFIG_SPEC, api={"port": 54})
+    {
+        "api": {
+            "authentication_token": "",
+            "hostname": "platform.strawberryfields.ai",
+            "use_ssl": True,
+            "port": 54,
+        }
+    }
+
+    Args:
+        config_spec (dict): nested dictionary representing the
+            configuration specification
+
+    Keyword Args:
+        Provided keyword arguments may overwrite default values of
+        matching keys.
+
+    Returns:
+        dict: the default configuration defined by the input config spec
+
+    Raises:
+        ConfigurationError: if provided keyword argument overrides do not
+        match the expected type defined in the configuration spec.
+    """
+    res = {}
+    for k, v in config_spec.items():
+        if isinstance(v, tuple):
+            # config spec value v represents the allowed type and default value
+
+            if k in kwargs:
+                # Key also exists as a keyword argument.
+                # Perform type validation.
+                if not isinstance(kwargs[k], v[0]):
+                    raise ConfigurationError(
+                        "Expected type {} for option {}, received {}".format(
+                            v[0], k, type(kwargs[k])
+                        )
+                    )
+
+                if kwargs[k] is not None:
+                    # Only add the key to the configuration object
+                    # if the provided override is not None.
+                    res[k] = kwargs[k]
+            else:
+                if v[1] is not None:
+                    # Only add the key to the configuration object
+                    # if the default value is not None.
+                    res[k] = v[1]
+
+        elif isinstance(v, dict):
+            # config spec value is a dictionary of more options
```

Comment:
(Minor suggestion): Just for consistency in names that we are using
```suggestion
            # config spec value is a configuration section
```